marimo/_runtime/runtime.py

```diff
@@ -588,6 +592,12 @@ def mutate_graph(
 
     def _run_cells(self, cell_ids: set[CellId_t]) -> None:
         """Run cells and any state updates they trigger"""
+
+        # Set __file__ to the globals
+        # we do this dynamically because the user may change the filename
+        # of the notebook
+        self.globals["__file__"] = self.app_metadata.filename
+
```

Comment:
Let's do this at kernel construction time, to match what the Python interpreter does (that way we'll also match the behavior when running as a script, e.g. if a user did something weird like overwriting `__file__`). It's okay for this to be stale if the user changes the filename.

Reply:
If we really want to update it at runtime, we can check to make sure the user hasn't written to `__file__` themselves. If they have, then don't update; otherwise, fine to update.

Reply:
made a change. i don't do it dynamically on each run, but i do set `__file__` whenever it gets changes via a Request